agera/src/main/java/com/google/android/agera/BaseObservable.java

```diff
@@ -38,149 +39,131 @@
  */
 public abstract class BaseObservable implements Observable {
   @NonNull
-  private final Worker worker;
+  private static final Object[] NO_UPDATABLES_OR_HANDLERS = new Object[0];
+  @NonNull
+  private final WorkerHandler handler;
+  private final int shortestUpdateWindowMillis;
+
+  @NonNull
+  private Object[] updatablesAndHandlers;
+  private int size;
+  private long lastUpdateTimestamp;
 
   protected BaseObservable() {
+    this(0);
+  }
+
+  BaseObservable(final int shortestUpdateWindowMillis) {
     checkState(Looper.myLooper() != null, "Can only be created on a Looper thread");
-    worker = new Worker(this);
+    this.shortestUpdateWindowMillis = shortestUpdateWindowMillis;
+    this.handler = workerHandler();
+    this.updatablesAndHandlers = NO_UPDATABLES_OR_HANDLERS;
+    this.size = 0;
   }
 
   @Override
-  public final void addUpdatable(@NonNull final Updatable updatable) {
+  public synchronized final void addUpdatable(@NonNull final Updatable updatable) {
     checkState(Looper.myLooper() != null, "Can only be added on a Looper thread");
-    worker.addUpdatable(updatable);
+    add(updatable, workerHandler());
+    if (size == 1) {
+      if (handler.hasMessages(MSG_LAST_REMOVED, this)) {
+        handler.removeMessages(MSG_LAST_REMOVED, this);
+      } else {
+        handler.obtainMessage(WorkerHandler.MSG_FIRST_ADDED, this).sendToTarget();
+      }
+    }
   }
 
   @Override
-  public final void removeUpdatable(@NonNull final Updatable updatable) {
+  public synchronized final void removeUpdatable(@NonNull final Updatable updatable) {
     checkState(Looper.myLooper() != null, "Can only be removed on a Looper thread");
```

Comment:
Same here, just moved but; yes, it's actually true. It keeps the references in a `ThreadLocal`, add/remove needs to be called paired from the same thread. This is also described in the javadoc for `Observable`